// idb-managed/build/db_wrapper.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @file Wrap idb APIs for idb-managed
 */
var utils_1 = require("./lib/utils");
var IDB = require('./lib/idb');
var IDB_MANAGER_VERSION = 1;
var IDB_MANAGER_DB_NAME = 'IDB_MANAGER_DB';
var IDB_MANAGER_DB_TABLE_NAME = 'IDB_MANAGER_STORE';
var IDB_MANAGER_DB_TABLE_INDEX_NAME = 'dbName';
var UPDATETIME_KEYNAME = 'updateTime';
var EXPIRETIME_KEYNAME = 'expireTime';
function indexRange2DBKey(indexRange) {
    var onlyIndex = indexRange.onlyIndex, lowerIndex = indexRange.lowerIndex, upperIndex = indexRange.upperIndex, _a = indexRange.lowerExclusive, lowerExclusive = _a === void 0 ? false : _a, _b = indexRange.upperExclusive, upperExclusive = _b === void 0 ? false : _b;
    if (onlyIndex !== undefined) {
        return IDBKeyRange.only(onlyIndex);
    }
    else if (lowerIndex !== undefined && upperIndex !== undefined) {
        return IDBKeyRange.bound(lowerIndex, upperIndex, lowerExclusive, upperExclusive);
    }
    else if (lowerIndex !== undefined) {
        return IDBKeyRange.lowerBound(lowerIndex, lowerExclusive);
    }
    else {
        return IDBKeyRange.upperBound(upperIndex, upperExclusive);
    }
}
function itemWrapper(itemConfig) {
    var _a;
    var currentTime = Date.now();
    return __assign(__assign({}, itemConfig.item), (_a = {}, _a[UPDATETIME_KEYNAME] = currentTime, _a[EXPIRETIME_KEYNAME] = itemConfig.itemDuration !== undefined
        ? itemConfig.itemDuration + currentTime
        : -1, _a));
}
function itemUnwrapper(item) {
    if (!item) {
        return null;
    }
    else if (item.expireTime > 0 && item.expireTime < Date.now()) {
        return null;
    }
    else {
        delete item.updateTime;
        delete item.expireTime;
        return item;
    }
}
function registerDBInManager(dbInfo) {
    return __awaiter(this, void 0, void 0, function () {
        var dbManager, dbAlreadyInManager, addDBTrans, table, dbItem;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, openDBManager()];
                case 1:
                    dbManager = _a.sent();
                    return [4 /*yield*/, getItemFromDB(dbManager, IDB_MANAGER_DB_TABLE_NAME, dbInfo.name)];
                case 2:
                    dbAlreadyInManager = (_a.sent());
                    if (!(!dbAlreadyInManager || dbInfo.version > dbAlreadyInManager.version)) return [3 /*break*/, 4];
                    addDBTrans = dbManager.transaction(IDB_MANAGER_DB_TABLE_NAME, 'readwrite');
                    table = addDBTrans.objectStore(IDB_MANAGER_DB_TABLE_NAME);
                    dbItem = {
                        dbName: dbInfo.name,
                        tableList: dbInfo.tableList,
                        version: dbInfo.version
                    };
                    table.put(itemWrapper({
                        item: dbItem,
                        tableName: IDB_MANAGER_DB_TABLE_NAME
                    }));
                    return [4 /*yield*/, addDBTrans.complete];
                case 3:
                    _a.sent();
                    dbManager.close();
                    _a.label = 4;
                case 4: return [2 /*return*/];
            }
        });
    });
}
function unregisterDBInManager(dbName) {
    return __awaiter(this, void 0, void 0, function () {
        var dbManager, deleteTrans, table;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, openDBManager()];
                case 1:
                    dbManager = _a.sent();
                    deleteTrans = dbManager.transaction(IDB_MANAGER_DB_TABLE_NAME, 'readwrite');
                    table = deleteTrans.objectStore(IDB_MANAGER_DB_TABLE_NAME);
                    table.delete(dbName);
                    return [4 /*yield*/, deleteTrans.complete];
                case 2:
                    _a.sent();
                    dbManager.close();
                    return [2 /*return*/];
            }
        });
    });
}
function createDB(dbInfo) {
    return __awaiter(this, void 0, void 0, function () {
        var db;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, registerDBInManager(dbInfo)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, IDB.open(dbInfo.name, dbInfo.version, function (upgradeDB) {
                            upgradeDBWithTableList(upgradeDB, dbInfo.tableList);
                        })];
                case 2:
                    db = _a.sent();
                    return [2 /*return*/, db];
            }
        });
    });
}
function openDBManager() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, IDB.open(IDB_MANAGER_DB_NAME, IDB_MANAGER_VERSION, 
                    // In case DB Manager has not been created.
                    function (upgradeDB) {
                        upgradeDBManager(upgradeDB);
                    })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
function openDB(dbName) {
    return __awaiter(this, void 0, void 0, function () {
        var dbManager, dbAlreadyInManager, db;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, openDBManager()];
                case 1:
                    dbManager = _a.sent();
                    return [4 /*yield*/, getItemFromDB(dbManager, IDB_MANAGER_DB_TABLE_NAME, dbName)];
                case 2:
                    dbAlreadyInManager = (_a.sent());
                    dbManager.close();
                    if (!dbAlreadyInManager) return [3 /*break*/, 4];
                    return [4 /*yield*/, IDB.open(dbAlreadyInManager.dbName, dbAlreadyInManager.version, 
                        // In case this DB has not been created.
                        function (upgradeDB) {
                            upgradeDBWithTableList(upgradeDB, dbAlreadyInManager.tableList || []);
                        })];
                case 3:
                    db = _a.sent();
                    return [2 /*return*/, db];
                case 4: return [2 /*return*/, null];
            }
        });
    });
}
function getItemFromDB(db, tableName, primaryKeyValue) {
    return __awaiter(this, void 0, void 0, function () {
        var trans, table, itemInTable;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!db.objectStoreNames.contains(tableName)) return [3 /*break*/, 2];
                    trans = db.transaction(tableName, 'readonly');
                    table = trans.objectStore(tableName);
                    return [4 /*yield*/, table.get(primaryKeyValue)];
                case 1:
                    itemInTable = (_a.sent());
                    return [2 /*return*/, itemUnwrapper(itemInTable)];
                case 2: return [2 /*return*/, null];
            }
        });
    });
}
function upgradeDBManager(upgradeDB) {
    upgradeDB.createObjectStore(IDB_MANAGER_DB_TABLE_NAME, {
        keyPath: IDB_MANAGER_DB_TABLE_INDEX_NAME
    });
}
function upgradeDBWithTableList(upgradeDB, tableList) {
    try {
        tableList.forEach(function (tableConfig) {
            // If table already exists.
            if (upgradeDB.objectStoreNames.contains(tableConfig.tableName)) {
                var currentTable_1 = upgradeDB
                    .transaction(tableConfig.tableName)
                    .objectStore(tableConfig.tableName);
                // Create new index for present table.
                (tableConfig.indexList || []).forEach(function (theIndex) {
                    if (!currentTable_1.indexNames.contains(theIndex.indexName)) {
                        currentTable_1.createIndex(theIndex.indexName, theIndex.indexName, { unique: theIndex.unique });
                    }
                });
                // Else create new table.
            }
            else {
                var tablePrimaryKey = tableConfig.primaryKey || 'id';
                var tableToCreate_1 = upgradeDB.createObjectStore(tableConfig.tableName, __assign({ keyPath: tablePrimaryKey }, (tablePrimaryKey === 'id'
                    ? { autoIncrement: true }
                    : {})));
                // Set index of primaryKey.
                tableToCreate_1.createIndex(tablePrimaryKey, tablePrimaryKey, {
                    unique: true
                });
                // Set indexes defined in tableConfig
                (tableConfig.indexList || []).forEach(function (theIndex) {
                    tableToCreate_1.createIndex(theIndex.indexName, theIndex.indexName, { unique: theIndex.unique });
                });
                // Set index of updateTime for data ordering priority.
                tableToCreate_1.createIndex(UPDATETIME_KEYNAME, UPDATETIME_KEYNAME, { unique: false });
                // Set index of expireTime for expired data deletion.
                tableToCreate_1.createIndex(EXPIRETIME_KEYNAME, EXPIRETIME_KEYNAME, { unique: false });
            }
        });
    }
    catch (e) {
        upgradeDB.close(); // Close upgraded DB to trigger the failure of this opening process.
    }
}
function atomicTrans(transaction, db, tryStatement) {
    return __awaiter(this, void 0, void 0, function () {
        var transError_1, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 3, 8, 9]);
                    return [4 /*yield*/, tryStatement()];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, transaction.complete];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 9];
                case 3:
                    transError_1 = _a.sent();
                    try {
                        // To roll back all operations in this transaction if any error happens.
                        transaction.abort();
                    }
                    catch (e) {
                        // Do nothing if transaction abort failed.
                    }
                    _a.label = 4;
                case 4:
                    _a.trys.push([4, 6, , 7]);
                    // To catch the Promise error caused by transaction abortion. Otherwise, uncaught rejection will be thrown up.
                    return [4 /*yield*/, transaction.complete];
                case 5:
                    // To catch the Promise error caused by transaction abortion. Otherwise, uncaught rejection will be thrown up.
                    _a.sent();
                    return [3 /*break*/, 7];
                case 6:
                    e_1 = _a.sent();
                    return [3 /*break*/, 7];
                case 7: 
                // Throw specific error happened in tryStatement.
                throw transError_1;
                case 8:
                    db.close();
                    return [7 /*endfinally*/];
                case 9: return [2 /*return*/];
            }
        });
    });
}
function deleteItemsFromDB(db, tableIndexRanges) {
    return __awaiter(this, void 0, void 0, function () {
        var validRanges, dedupTableNameList, deleteItemsTrans;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    validRanges = tableIndexRanges.filter(function (indexRange) {
                        return db.objectStoreNames.contains(indexRange.tableName);
                    });
                    dedupTableNameList = utils_1.deduplicateList(validRanges.map(function (tableIndexRange) { return tableIndexRange.tableName; }));
                    deleteItemsTrans = db.transaction(dedupTableNameList, 'readwrite');
                    return [4 /*yield*/, atomicTrans(deleteItemsTrans, db, function () { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, Promise.all(validRanges.map(function (tableIndexRange) {
                                            var tableName = tableIndexRange.tableName, indexRange = tableIndexRange.indexRange;
                                            var table = deleteItemsTrans.objectStore(tableName);
                                            if (!indexRange) {
                                                return table.clear();
                                            }
                                            else {
                                                return new Promise(function (resolve) {
                                                    table.index(indexRange.indexName).iterateCursor(indexRange2DBKey(indexRange), function (cursor) {
                                                        if (!cursor) {
                                                            resolve();
                                                            return;
                                                        }
                                                        table.delete(cursor.primaryKey);
                                                        cursor.continue();
                                                    });
                                                });
                                            }
                                        }))];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function addItems(dbInfo, items) {
    return __awaiter(this, void 0, void 0, function () {
        var dedupTableNameList, db, addItemsTrans;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    dedupTableNameList = utils_1.deduplicateList(items.map(function (item) { return item.tableName; }));
                    return [4 /*yield*/, deleteItems(dbInfo.name, dedupTableNameList.map(function (tableName) {
                            return {
                                tableName: tableName,
                                indexRange: {
                                    indexName: EXPIRETIME_KEYNAME,
                                    upperIndex: +new Date(),
                                    upperExclusive: false
                                }
                            };
                        }))];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, createDB(dbInfo)];
                case 2:
                    db = _a.sent();
                    addItemsTrans = db.transaction(dedupTableNameList, 'readwrite');
                    return [4 /*yield*/, atomicTrans(addItemsTrans, db, function () { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, Promise.all(items.map(function (item) { return addItemsTrans.objectStore(item.tableName).put(itemWrapper(item)); }))];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 3:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.addItems = addItems;
function getItem(dbName, tableName, primaryKeyValue) {
    return __awaiter(this, void 0, void 0, function () {
        var db, item, e_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, openDB(dbName)];
                case 1:
                    db = _a.sent();
                    if (!db) return [3 /*break*/, 7];
                    _a.label = 2;
                case 2:
                    _a.trys.push([2, 4, 5, 6]);
                    return [4 /*yield*/, getItemFromDB(db, tableName, primaryKeyValue)];
                case 3:
                    item = _a.sent();
                    return [2 /*return*/, item];
                case 4:
                    e_2 = _a.sent();
                    throw e_2;
                case 5:
                    db.close();
                    return [7 /*endfinally*/];
                case 6: return [3 /*break*/, 8];
                case 7: return [2 /*return*/, null];
                case 8: return [2 /*return*/];
            }
        });
    });
}
exports.getItem = getItem;
function getItemsInRange(dbName, tableIndexRange) {
    return __awaiter(this, void 0, void 0, function () {
        var tableName, indexRange, db, items_1, trans, table_1, wrappedItems, e_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    tableName = tableIndexRange.tableName, indexRange = tableIndexRange.indexRange;
                    return [4 /*yield*/, openDB(dbName)];
                case 1:
                    db = _a.sent();
                    if (!db) return [3 /*break*/, 11];
                    _a.label = 2;
                case 2:
                    _a.trys.push([2, 8, 9, 10]);
                    items_1 = [];
                    if (!!db.objectStoreNames.contains(tableName)) return [3 /*break*/, 3];
                    return [3 /*break*/, 7];
                case 3:
                    trans = db.transaction(tableName, 'readonly');
                    table_1 = trans.objectStore(tableName);
                    if (!!indexRange) return [3 /*break*/, 5];
                    return [4 /*yield*/, table_1.getAll()];
                case 4:
                    wrappedItems = _a.sent();
                    items_1 = (wrappedItems || [])
                        .map(itemUnwrapper)
                        .filter(function (item) {
                        return item !== null;
                    });
                    return [3 /*break*/, 7];
                case 5: return [4 /*yield*/, new Promise(function (resolve) {
                        table_1.index(indexRange.indexName).iterateCursor(indexRange2DBKey(indexRange), function (cursor) {
                            if (!cursor) {
                                resolve();
                                return;
                            }
                            var item = itemUnwrapper(cursor.value);
                            item && items_1.push(item);
                            cursor.continue();
                        });
                    })];
                case 6:
                    _a.sent();
                    _a.label = 7;
                case 7: return [2 /*return*/, items_1];
                case 8:
                    e_3 = _a.sent();
                    throw e_3;
                case 9:
                    db.close();
                    return [7 /*endfinally*/];
                case 10: return [3 /*break*/, 12];
                case 11: return [2 /*return*/, []];
                case 12: return [2 /*return*/];
            }
        });
    });
}
exports.getItemsInRange = getItemsInRange;
function deleteDB(dbName) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, unregisterDBInManager(dbName)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, IDB.delete(dbName)];
                case 2:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.deleteDB = deleteDB;
function deleteItems(dbName, tableIndexRanges) {
    return __awaiter(this, void 0, void 0, function () {
        var db;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, openDB(dbName)];
                case 1:
                    db = _a.sent();
                    if (!db) return [3 /*break*/, 3];
                    return [4 /*yield*/, deleteItemsFromDB(db, tableIndexRanges)];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 4];
                case 3: 
                // If db does not exist, no need to deleteItems at all.
                return [2 /*return*/];
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.deleteItems = deleteItems;
exports.default = {
    addItems: addItems,
    getItem: getItem,
    getItemsInRange: getItemsInRange,
    deleteDB: deleteDB,
    deleteItems: deleteItems
};
